### ISSUE TITLE
perf: add pin_memory, num_workers, and mixed precision to custom training loops

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#6482]
## Problem

The custom PyTorch training loops in `scripts/training/` miss standard GPU training optimizations, leaving performance on the table.

### 1. DataLoader missing `pin_memory` and `num_workers` (Medium)

**File:** `scripts/training/train_predictive_planner.py:339-340`

```python
train_loader = DataLoader(_ds(train_idx), batch_size=batch_size, shuffle=True, drop_last=False)
```

No `pin_memory=True`, `num_workers`, or `prefetch_factor`. Every batch transfer goes through pageable memory, and data loading is synchronous on the main process. The training loop then does four separate synchronous `.to(device)` calls per batch.

**Fix:** Add `pin_memory=True` (when CUDA), `num_workers=2-4`, `persistent_workers=True`, and `non_blocking=True` on `.to(device)` calls.

### 2. No mixed precision training anywhere (Medium)

**Files:** All training scripts under `scripts/training/` and `robot_sf/training/`

Zero matches for `torch.cuda.amp`, `GradScaler`, `torch.autocast`, or `torch.amp`. All custom training loops run in FP32. Mixed precision typically provides 1.5-3x throughput on modern NVIDIA GPUs.

**Fix:** Wrap forward/backward in `torch.autocast(device_type="cuda", dtype=torch.float16)` with `GradScaler` for custom training loops.

### 3. O(n) replay buffer eviction via `list.pop(0)` (Low-Medium)

**File:** `scripts/training/train_distributional_rl.py:432-441`

```python
if len(replay) > config.dqn.replay_size:
    replay.pop(0)  # O(n) shift of ~4096 elements
```

**Fix:** Replace with `collections.deque(maxlen=config.dqn.replay_size)`.

## Acceptance criteria

- [ ] `train_predictive_planner.py` DataLoader uses `pin_memory`, `num_workers`, and `non_blocking` transfers.
- [ ] At least one custom training loop demonstrates AMP usage.
- [ ] Replay buffer uses O(1) eviction.
- [ ] Training produces equivalent results (loss curves within noise).

## Evidence

Code review on 2026-07-30.

**Comment by ll7** [comment]
Audit application — first bounded child: #6488 owns only the predictive-planner DataLoader worker/pinned-memory transfer path. AMP and replay-buffer work remain separate. Duplicate search found no equivalent issue or PR. Decision source: Codex task 019fb13c-6ea8-7632-a41b-f9940898a48c.

**Comment by ll7** [comment]
factory-v2: split into bounded child issues; parent remains coordination-only

**Comment by ll7** [comment]
### Blocked-queue triage (tracking #7067)



```yaml
blocker_class: upstream_issue
unblock_condition: >-
  The bounded children covering the DataLoader transfer path (#6488), AMP/mixed precision, and replay-buffer work are all closed.
watcher: >-
  `gh issue list --search 'is:open 6482 in:body'` returns no open children
next_check_at: >-
  2026-09-01
last_meaningful_progress_at: "2026-07-30"
```

Converted to coordination-only on 2026-07-30; the audit named #6488 as the first bounded child.

Bookkeeping only — no scope, claim, label, or evidence change is made by this comment, and the issue stays open.

**Comment by ll7** [comment]
Current-main audit for #6482 at `origin/main=4aac85935985434cf25c25442a1bc5bc216b78e3`:

- The replay-buffer slice is already implemented by `f216d2f4810b9d2176043f84284053ad5e1464c1`
  (`perf(training): use bounded deque for O(1) replay eviction`, merged via PR #6524).
- `scripts/training/train_distributional_rl.py` now uses `deque(maxlen=config.dqn.replay_size)`;
  the dedicated contract covers FIFO eviction, `random.sample` compatibility, trainer sampling,
  and training with eviction active.
- Linked-worktree proof at this current base: `tests/training/test_train_distributional_rl_replay_buffer.py`
  passed 4 tests.

This does not close or relabel #6482. The DataLoader/pinned-memory worker criterion and AMP
criterion remain open and need separate implementation and equivalence proof; no benchmark speedup,
GPU result, or completion claim is made. Existing child #6488 covers the predictive-planner
DataLoader direction; the parent remains coordination-only until the remaining criteria are handled.

**Comment by ll7** [comment]
### Corrective current-main audit

This supersedes the earlier current-main note in comment `5287039096`, which incorrectly
described the DataLoader and AMP child work as still open.

At `origin/main=a956b839e877ce18d03a5aa7903ff923952547da`, all three bounded implementation
slices for this coordination issue are present and their child issues are closed:

- DataLoader worker/pinned-memory path: #6488, implemented by merged PR #6518
  (`156cb30990fafd667ae839e2543b1d75dc448742`), including effective-setting manifest
  provenance and CPU-safe/non-blocking gating.
- CUDA-gated opt-in AMP path: #6509, implemented by merged PR #6523
  (`14b82d452919593145f0545445a4b9da0686cadc`), with the default CPU path left unchanged.
- Bounded replay eviction: #6508, implemented by merged PR #6524
  (`f216d2f4810b9d2176043f84284053ad5e1464c1`), using FIFO `deque(maxlen=...)` while
  preserving sampling and the diagnostic CPU trajectory contract.

This correction records implementation state only. It does not claim a GPU speedup, a complete
end-to-end performance result, or a new equivalence result for the parent acceptance criterion.
Those claims still require the appropriate measured evidence. #6482 remains open and
coordination-only; no close or label change is requested by this note.

**Comment by ll7** [comment]
## Unblock route — 2026-08-16

Split the work by evidence risk. **Ready now:** replay-buffer/deque and DataLoader efficiency changes with deterministic equivalence, ordering, sampling, and seed tests on CPU fixtures. **Separately gated:** AMP/mixed-precision changes require GPU parity checks for losses, gradients, checkpoints, and evaluation outputs before training use.

Merge the low-risk data-path slice first. Do not make AMP the blocker for the entire issue.

**Comment by ll7** [comment]
## Readiness reconciliation — 2026-08-17

All three implementation slices are merged and their child issues are closed. The remaining parent acceptance criterion is empirical: show that the loader/AMP paths preserve a bounded short-run training signal while recording the actual throughput and environment.

That proof is now isolated in #7254. The repository-wide unblock pass incorrectly marked this coordination parent as locally ready. The parent is restored to `state:blocked` with `resource:gpu`; dispatch #7254 rather than changing the training loops again. No benchmark, policy-quality, or general GPU-speed claim is authorized.

**Comment by ll7** [comment]
## Child #7254 evidence: bounded GPU optimization smoke

Child #7254 completed on `b2576e8ad55c50dcfd04aca74ab1bc9de805287b` using the frozen config-first comparison.

- Two repeats each of FP32 control, FP32 loader optimization, and AMP + loader optimization.
- RTX 3080 / PyTorch `2.13.0+cu130` / CUDA `13.0`.
- All runs were finite, processed identical `512 x 3 x 4` fixture rows and update counts, and produced readable checkpoints.
- Both optimized arms were within the predeclared `1e-3` curve envelope: `equivalent_smoke`.
- Post-warm-up throughput on this tiny fixture was lower than control: control `33,712` examples/s, loader `27,296`, AMP+loader `15,761`.
- Peak allocated CUDA memory: control/loader `17,415,168` bytes; AMP `17,344,000` bytes.

Conclusion: the implementation-level equivalence criterion is satisfied for this fixture, but the observed throughput does not justify a general speedup/default recommendation. This remains diagnostic smoke evidence, not policy-quality or benchmark evidence. The reusable runner/config and per-epoch provenance are being delivered through the child implementation PR.

**Comment by ll7** [comment]
## Child #7254 evidence: bounded GPU optimization smoke (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on
`c07d74d38323ef115fd9c123766e0f852a7312e8` using the frozen config-first comparison.

- Two repeats each of FP32 control, FP32 loader optimization, and AMP + loader optimization.
- RTX 3080 / PyTorch `2.13.0+cu130` / CUDA `13.0`.
- All runs were finite, processed identical `512 x 3 x 4` fixture rows and update counts, and produced readable checkpoints.
- Both optimized arms were within the predeclared `1e-3` curve envelope: `equivalent_smoke`.
- Post-warm-up throughput on this tiny fixture was lower than control: control `35,756` examples/s, loader `26,806`, AMP+loader `21,738`.
- Peak allocated CUDA memory: control/loader `17,415,168` bytes; AMP `17,344,000` bytes.

Conclusion: the implementation-level equivalence criterion is satisfied for this fixture, but the observed throughput does not justify a general speedup/default recommendation. This remains diagnostic smoke evidence, not policy-quality or benchmark evidence. The reusable runner/config and per-epoch provenance are being delivered through child PR #7263.

**Comment by ll7** [comment]
## Child #7254 evidence: bounded GPU optimization smoke (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on
`6a395b39819764d98a6c3c7beca4b8bbac7bd1fc` using the frozen config-first comparison after
refreshing the branch onto live `origin/main`.

- Two repeats each of FP32 control, FP32 loader optimization, and AMP + loader optimization.
- RTX 3080 / PyTorch `2.13.0+cu130` / CUDA `13.0`.
- All runs were finite, processed identical `512 x 3 x 4` fixture rows and update counts, and produced readable checkpoints.
- Both optimized arms were within the predeclared `1e-3` curve envelope: `equivalent_smoke`.
- Post-warm-up throughput on this tiny fixture was lower than control for FP32 loader: control `33,999` examples/s, loader `22,818`, AMP+loader `23,246`; throughput is noisy on this tiny fixture and remains diagnostic only.
- Peak allocated CUDA memory: control/loader `17,415,168` bytes; AMP `17,344,000` bytes.

Conclusion: the implementation-level equivalence criterion is satisfied for this fixture, but the observed throughput does not justify a general speedup/default recommendation. This remains diagnostic smoke evidence, not policy-quality or benchmark evidence. The reusable runner/config and per-epoch provenance are being delivered through child PR #7263.

**Comment by ll7** [comment]
## Child #7254 evidence: bounded GPU optimization smoke (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on
`36abcad38c8bf4e6bec653c07547dffe43c8a4c0` using the frozen config-first comparison after
refreshing the branch onto live `origin/main` and adding the fail-closed summary contract.

- Two repeats each of FP32 control, FP32 loader optimization, and AMP + loader optimization.
- RTX 3080 / PyTorch `2.13.0+cu130` / CUDA `13.0`.
- All runs were finite, processed identical `512 x 3 x 4` fixture rows and update counts, and produced readable checkpoints.
- Both optimized arms were within the predeclared `1e-3` curve envelope: `equivalent_smoke`.
- Post-warm-up throughput on this tiny fixture was lower than control: control `32,842` examples/s, loader `23,109`, AMP+loader `18,960`; throughput is noisy on this tiny fixture and remains diagnostic only.
- Peak allocated CUDA memory: control/loader `17,415,168` bytes; AMP `17,344,000` bytes.

Conclusion: the implementation-level equivalence criterion is satisfied for this fixture, but the observed throughput does not justify a general speedup/default recommendation. This remains diagnostic smoke evidence, not policy-quality or benchmark evidence. The reusable runner/config and per-epoch provenance are being delivered through child PR #7263.

**Comment by ll7** [comment]
## Child #7254 evidence: bounded GPU optimization smoke (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on
`33e1aabbb778a3ef4cbe9dc379c0b3ced990fc9a` using the frozen config-first comparison after
refreshing the branch onto live `origin/main` and adding the fail-closed summary contract.

- Two repeats each of FP32 control, FP32 loader optimization, and AMP + loader optimization.
- RTX 3080 / PyTorch `2.13.0+cu130` / CUDA `13.0`.
- All runs were finite, processed identical `512 x 3 x 4` fixture rows and update counts, and produced readable checkpoints.
- Both optimized arms were within the predeclared `1e-3` curve envelope: `equivalent_smoke`.
- Post-warm-up throughput on this tiny fixture was lower than control: control `32,500` examples/s, loader `24,106`, AMP+loader `18,649`; throughput is noisy on this tiny fixture and remains diagnostic only.
- Peak allocated CUDA memory: control/loader `17,415,168` bytes; AMP `17,344,000` bytes.

Conclusion: the implementation-level equivalence criterion is satisfied for this fixture, but the observed throughput does not justify a general speedup/default recommendation. This remains diagnostic smoke evidence, not policy-quality or benchmark evidence. The reusable runner/config and per-epoch provenance are being delivered through child PR #7263.

**Comment by ll7** [comment]
## Parent completion audit after #7254

All four acceptance criteria are now satisfied:

- DataLoader worker/pinned-memory/non-blocking path: merged implementation slices #6488/#6518.
- Custom-loop AMP path: merged #6509/#6523.
- O(1) replay eviction: merged #6508/#6524.
- Matched training-signal equivalence: completed child #7254 / merged PR #7263.

The final exact-head RTX 3080 fixture result was `equivalent_smoke`: control `32,500`, FP32 loader `24,106`, and AMP+loader `18,649` post-warm-up examples/s, with both optimized arms inside the frozen `1e-3` curve envelope. This is diagnostic one-fixture evidence only; it does not justify a default or general speedup claim.

The remaining empirical question is deliberately split out to [#7290](https://github.com/ll7/robot_sf_ll7/issues/7290): qualify throughput on one realistic canonical dataset before any default recommendation. Closing this parent records that its implementation acceptance criteria are complete without overstating the performance result.

**Comment by ll7** [comment]
## Parent acceptance reconciliation

Parent issue #6482 is reopened pending review of the corrected evidence child [#7254](https://github.com/ll7/robot_sf_ll7/issues/7254) and replacement [PR #7364](https://github.com/ll7/robot_sf_ll7/pull/7364) at exact head `5bd1f0d29fb19592bd47d61bda487670ee6438fc`.

The child now supplies a tracked compact manifest at [`docs/context/evidence/issue_7254_predictive_optimization_smoke_2026-08-17/summary.json`](https://github.com/ll7/robot_sf_ll7/blob/5bd1f0d29fb19592bd47d61bda487670ee6438fc/docs/context/evidence/issue_7254_predictive_optimization_smoke_2026-08-17/summary.json) with matched sampler order, dataset/split/model/optimizer/update-count identity, strict checkpoint loads, a frozen comparison envelope, and environment provenance. It supports bounded one-fixture loss-curve equivalence only.

The observed optimized-arm throughput is lower on this run (21,239.02 FP32-loader and 22,104.56 AMP+loader versus 33,678.22 control), so no default-performance recommendation follows. Keep the parent acceptance criterion and issue open until independent review and hosted checks confirm the evidence boundary; do not treat this diagnostic smoke as a general performance result.

**Comment by ll7** [comment]
Completed following child #7254 and PR #7364 (commit `75387dd4d`):

- DataLoader pinned memory, worker configuration, and non-blocking transfers are implemented.
- Custom training loops support AMP with GradScaler.
- Replay buffer eviction is O(1).
- Corrected diagnostic optimization smoke evidence is recorded in `docs/context/evidence/issue_7254_predictive_optimization_smoke_2026-08-17/summary.json`.
